File: packages/js/src/nlp/contextAnalyzer.ts
// src/nlp/contextAnalyzer.ts
import { Language } from '../types/types';

export interface ContextAnalysisResult {
  contextScore: number; // 0-1, where 0 = negative context, 1 = positive context
  reason: string;
  isWhitelisted: boolean;
}

export interface ContextConfig {
  contextWindow: number;
  language: Language;
  domainWhitelists?: string[];
}

// Positive sentiment indicators
const POSITIVE_INDICATORS = new Set([
  'amazing', 'awesome', 'excellent', 'fantastic', 'great', 'love', 'wonderful',
  'brilliant', 'perfect', 'incredible', 'outstanding', 'superb', 'magnificent',
  'marvelous', 'spectacular', 'phenomenal', 'terrific', 'fabulous', 'divine',
  'best', 'good', 'nice', 'cool', 'sweet', 'rad', 'sick', 'dope', 'fire',
  'lit', 'epic', 'legendary', 'godlike', 'insane', 'crazy', 'wild', 'beast',
  'movie', 'film', 'show', 'song', 'music', 'game', 'book', 'restaurant',
  'food', 'dish', 'meal', 'place', 'spot', 'location', 'experience'
]);

// Negative sentiment indicators
const NEGATIVE_INDICATORS = new Set([
  'hate', 'terrible', 'awful', 'horrible', 'disgusting', 'pathetic', 'stupid',
  'idiot', 'moron', 'loser', 'worthless', 'useless', 'garbage', 'trash',
  'suck', 'sucks', 'worst', 'bad', 'ugly', 'gross', 'nasty', 'annoying',
  'irritating', 'frustrating', 'disappointing', 'lame', 'weak', 'fail',
  'you', 'your', 'yourself', 'u', 'ur', 'ure', 'youre'
]);


// Domain-specific positive contexts
const GAMING_POSITIVE = new Set([
  'player', 'gamer', 'team', 'squad', 'clan', 'guild', 'match', 'game',
  'round', 'level', 'boss', 'raid', 'quest', 'achievement', 'skill',
  'build', 'loadout', 'strategy', 'tactic', 'play', 'move', 'combo'
]);

// Words that are acceptable in gaming contexts but might be flagged otherwise
const GAMING_ACCEPTABLE_WORDS = new Set([
  'kill', 'killer', 'killed', 'killing',
  'shoot', 'shot', 'shooting',
  'die', 'dying', 'died', 'dead', 'death',
  'badass', 'sick', 'insane', 'crazy', 'mad', 'beast', 'savage',
  'suck', 'sucks',
  'wtf', 'omg', 'hell', 'damn', 'crap'
]);

// Common positive phrases that might contain flagged words
const POSITIVE_PHRASES = new Map([
  ['the bomb', 0.9], // "this movie is the bomb"
  ['da bomb', 0.9], // slang for "the best"
  ['bomb.com', 0.9], // website reference
  ['bomb diggity', 0.9], // slang for excellent
  ['photo bomb', 0.8], // photography term
  ['bath bomb', 0.8], // cosmetic product
  ['bomb squad', 0.7], // could be neutral/positive in gaming
]);

// Negative phrases that should remain flagged
const NEGATIVE_PHRASES = new Map([
  ['you are', 0.1], // "you are [profanity]"
  ['ur a', 0.1], // "ur a [profanity]"
  ['such a', 0.2], // "such a [profanity]"
  ['fucking', 0.1], // intensifier, usually negative
  ['damn', 0.2], // mild profanity, context dependent
]);

export class ContextAnalyzer {
  private contextWindow: number;
  private language: Language;
  private domainWhitelists: Set<string>;

  constructor(config: ContextConfig) {
    this.contextWindow = config.contextWindow;
    this.language = config.language;
    this.domainWhitelists = new Set(config.domainWhitelists || []);
  }

  /**
   * Analyzes the context around a profanity match to determine if it should be flagged
   */
  analyzeContext(
    text: string,
    matchWord: string,
    matchIndex: number
  ): ContextAnalysisResult {
    const words = this.tokenize(text);
    const matchWordIndex = this.findWordIndex(words, matchIndex);
    
    if (matchWordIndex === -1) {
      return {
        contextScore: 0.5,
        reason: 'Could not locate match in tokenized text',
        isWhitelisted: false
      };
    }

    // Extract context window
    const startIndex = Math.max(0, matchWordIndex - this.contextWindow);
    const endIndex = Math.min(words.length, matchWordIndex + this.contextWindow + 1);
    const contextWords = words.slice(startIndex, endIndex);
    const contextText = contextWords.join(' ').toLowerCase();

    // Check for exact phrase matches first
    const phraseResult = this.checkPhraseContext(contextText, matchWord);
    if (phraseResult) {
      return phraseResult;
    }

    // Check domain-specific whitelists
    if (this.isDomainWhitelisted(contextWords, matchWord)) {
      return {
        contextScore: 0.8,
        reason: 'Domain-specific whitelist match',
        isWhitelisted: true
      };
    }

    // Perform sentiment analysis
    const sentimentScore = this.calculateSentimentScore(contextWords, matchWordIndex - startIndex);
    
    return {
      contextScore: sentimentScore,
      reason: this.generateReason(sentimentScore, contextWords),
      isWhitelisted: false
    };
  }

  private checkPhraseContext(contextText: string, matchWord: string): ContextAnalysisResult | null {
    // Check positive phrases
    for (const [phrase, score] of POSITIVE_PHRASES.entries()) {
      if (phrase.includes(matchWord) && contextText.includes(phrase)) {
        return {
          contextScore: score,
          reason: `Positive phrase detected: "${phrase}"`,
          isWhitelisted: true
        };
      }
    }

    // Check negative phrases (prefixes like "you are" that introduce profanity)
    for (const [phrase, score] of NEGATIVE_PHRASES.entries()) {
      if (contextText.includes(phrase)) {
        return {
          contextScore: score,
          reason: `Negative phrase detected: "${phrase}"`,
          isWhitelisted: false
        };
      }
    }

    return null;
  }

  private isDomainWhitelisted(contextWords: string[], matchWord: string): boolean {
    const normalizedMatchWord = matchWord.toLowerCase();

    // Check if any domain whitelist words are present
    for (const word of contextWords) {
      // Check user-defined domain whitelists (permissive)
      if (this.domainWhitelists.has(word)) {
        return true;
      }

      // Check internal gaming whitelist (restrictive)
      if (GAMING_POSITIVE.has(word)) {
        if (GAMING_ACCEPTABLE_WORDS.has(normalizedMatchWord)) {
          return true;
        }
      }
    }
    return false;
  }

  private generateReason(score: number, contextWords: string[]): string {
    const foundPositive = Array.from(new Set(contextWords.filter(word => POSITIVE_INDICATORS.has(word))));
    const foundNegative = Array.from(new Set(contextWords.filter(word => NEGATIVE_INDICATORS.has(word))));

    if (score >= 0.7) {
      const details = foundPositive.length > 0 ? ` (found: ${foundPositive.join(', ')})` : '';
      return `Positive context detected${details} - likely not profanity`;
    } else if (score <= 0.3) {
      const details = foundNegative.length > 0 ? ` (found: ${foundNegative.join(', ')})` : '';
      return `Negative context detected${details} - likely profanity`;
    } else {
      return 'Neutral context - uncertain classification';
    }
  }

  private tokenize(text: string): string[] {
    // Simple tokenization - split on whitespace and punctuation
    return text.toLowerCase()
      .replace(/[^\w\s]/g, ' ')
      .split(/\s+/)
      .filter(word => word.length > 0);
  }

  private findWordIndex(words: string[], charIndex: number): number {
    // This is a simplified approach - in production, you'd want more robust mapping
    // For now, we'll estimate based on the character position
    let currentPos = 0;
    for (let i = 0; i < words.length; i++) {
      if (currentPos >= charIndex) {
        return Math.max(0, i - 1);
      }
      currentPos += words[i].length + 1; // +1 for space
    }
    return words.length - 1;
  }


  private calculateSentimentScore(contextWords: string[], matchPosition: number): number {
    let positiveCount = 0;
    let negativeCount = 0;
    const totalWords = contextWords.length;

    // Weight words closer to the match more heavily
    for (let i = 0; i < contextWords.length; i++) {
      const word = contextWords[i];
      const distance = Math.abs(i - matchPosition);
      const weight = Math.max(0.1, 1 - (distance * 0.2)); // Closer words have higher weight

      if (POSITIVE_INDICATORS.has(word)) {
        positiveCount += weight;
      } else if (NEGATIVE_INDICATORS.has(word)) {
        negativeCount += weight;
      }
    }

    // Calculate base score
    const totalSentiment = positiveCount + negativeCount;
    if (totalSentiment === 0) {
      return 0.5; // Neutral if no sentiment indicators
    }

    const rawScore = positiveCount / totalSentiment;
    
    // Apply context-specific adjustments
    let adjustedScore = rawScore;
    
    // Adjust confidence based on context window size
    const confidenceMultiplier = Math.min(1.0, totalWords / 5); // More words = higher confidence
    adjustedScore = 0.5 + (adjustedScore - 0.5) * confidenceMultiplier;

    // If there are personal pronouns (you, your), lean towards negative
    const hasPersonalPronouns = contextWords.some(word => 
      ['you', 'your', 'u', 'ur'].includes(word)
    );
    if (hasPersonalPronouns && rawScore < 0.7) {
      adjustedScore *= 0.7; // Reduce score when personal pronouns are present
    }

    // If there are object/thing references, lean towards positive
    const hasObjectReferences = contextWords.some(word =>
      ['movie', 'song', 'game', 'book', 'show', 'this', 'that', 'it'].includes(word)
    );
    if (hasObjectReferences && rawScore > 0.3) {
      adjustedScore = Math.min(1, adjustedScore * 1.3); // Boost score for object references
    }

    return Math.max(0, Math.min(1, adjustedScore));
  }


  /**
   * Updates the domain whitelist for this analyzer instance
   */
  updateDomainWhitelist(newWhitelist: string[]): void {
    this.domainWhitelists = new Set(newWhitelist);
  }

  /**
   * Adds words to the domain whitelist
   */
  addToDomainWhitelist(words: string[]): void {
    words.forEach(word => this.domainWhitelists.add(word.toLowerCase()));
  }
}

File: packages/py/glin_profanity/nlp/context_analyzer.py
"""Context analysis for distinguishing profanity from false positives."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

Language = Literal[
    "arabic",
    "chinese",
    "czech",
    "danish",
    "dutch",
    "english",
    "esperanto",
    "finnish",
    "french",
    "german",
    "hindi",
    "hungarian",
    "italian",
    "japanese",
    "korean",
    "norwegian",
    "persian",
    "polish",
    "portuguese",
    "russian",
    "spanish",
    "swedish",
    "thai",
    "turkish",
]


@dataclass(frozen=True)
class ContextAnalysisResult:
    context_score: float
    reason: str
    is_whitelisted: bool


@dataclass
class ContextConfig:
    context_window: int
    language: Language
    domain_whitelists: list[str] | None = None


POSITIVE_INDICATORS = frozenset(
    {
        "amazing",
        "awesome",
        "excellent",
        "fantastic",
        "great",
        "love",
        "wonderful",
        "brilliant",
        "perfect",
        "incredible",
        "outstanding",
        "superb",
        "magnificent",
        "marvelous",
        "spectacular",
        "phenomenal",
        "terrific",
        "fabulous",
        "divine",
        "best",
        "good",
        "nice",
        "cool",
        "sweet",
        "rad",
        "sick",
        "dope",
        "fire",
        "lit",
        "epic",
        "legendary",
        "godlike",
        "insane",
        "crazy",
        "wild",
        "beast",
        "movie",
        "film",
        "show",
        "song",
        "music",
        "game",
        "book",
        "restaurant",
        "food",
        "dish",
        "meal",
        "place",
        "spot",
        "location",
        "experience",
    }
)

NEGATIVE_INDICATORS = frozenset(
    {
        "hate",
        "terrible",
        "awful",
        "horrible",
        "disgusting",
        "pathetic",
        "stupid",
        "idiot",
        "moron",
        "loser",
        "worthless",
        "useless",
        "garbage",
        "trash",
        "suck",
        "sucks",
        "worst",
        "bad",
        "ugly",
        "gross",
        "nasty",
        "annoying",
        "irritating",
        "frustrating",
        "disappointing",
        "lame",
        "weak",
        "fail",
        "you",
        "your",
        "yourself",
        "u",
        "ur",
        "ure",
        "youre",
    }
)

GAMING_POSITIVE = frozenset(
    {
        "player",
        "gamer",
        "team",
        "squad",
        "clan",
        "guild",
        "match",
        "game",
        "round",
        "level",
        "boss",
        "raid",
        "quest",
        "achievement",
        "skill",
        "build",
        "loadout",
        "strategy",
        "tactic",
        "play",
        "move",
        "combo",
    }
)

GAMING_ACCEPTABLE_WORDS = frozenset(
    {
        "kill",
        "killer",
        "killed",
        "killing",
        "shoot",
        "shot",
        "shooting",
        "die",
        "dying",
        "died",
        "dead",
        "death",
        "badass",
        "sick",
        "insane",
        "crazy",
        "mad",
        "beast",
        "savage",
        "suck",
        "sucks",
        "wtf",
        "omg",
        "hell",
        "damn",
        "crap",
    }
)

POSITIVE_PHRASES: dict[str, float] = {
    "the bomb": 0.9,
    "da bomb": 0.9,
    "bomb.com": 0.9,
    "bomb diggity": 0.9,
    "photo bomb": 0.8,
    "bath bomb": 0.8,
    "bomb squad": 0.7,
}

NEGATIVE_PHRASES: dict[str, float] = {
    "you are": 0.1,
    "ur a": 0.1,
    "such a": 0.2,
    "fucking": 0.1,
    "damn": 0.2,
}


class ContextAnalyzer:
    """Analyzes surrounding text to reduce context-dependent false positives."""

    def __init__(self, config: ContextConfig) -> None:
        self.context_window = config.context_window
        self.language = config.language
        self.domain_whitelists = {
            word.lower() for word in (config.domain_whitelists or [])
        }

    def analyze_context(
        self, text: str, match_word: str, match_index: int
    ) -> ContextAnalysisResult:
        words = self._tokenize(text)
        match_word_index = self._find_word_index(words, match_index)

        if match_word_index == -1:
            return ContextAnalysisResult(
                context_score=0.5,
                reason="Could not locate match in tokenized text",
                is_whitelisted=False,
            )

        start_index = max(0, match_word_index - self.context_window)
        end_index = min(len(words), match_word_index + self.context_window + 1)
        context_words = words[start_index:end_index]
        context_text = " ".join(context_words).lower()

        phrase_result = self._check_phrase_context(context_text, match_word)
        if phrase_result is not None:
            return phrase_result

        if self._is_domain_whitelisted(context_words, match_word):
            return ContextAnalysisResult(
                context_score=0.8,
                reason="Domain-specific whitelist match",
                is_whitelisted=True,
            )

        sentiment_score = self._calculate_sentiment_score(
            context_words, match_word_index - start_index
        )
        return ContextAnalysisResult(
            context_score=sentiment_score,
            reason=self._generate_reason(sentiment_score, context_words),
            is_whitelisted=False,
        )

    def update_domain_whitelist(self, new_whitelist: list[str]) -> None:
        self.domain_whitelists = {word.lower() for word in new_whitelist}

    def add_to_domain_whitelist(self, words: list[str]) -> None:
        self.domain_whitelists.update(word.lower() for word in words)

    def _check_phrase_context(
        self, context_text: str, match_word: str
    ) -> ContextAnalysisResult | None:
        for phrase, score in POSITIVE_PHRASES.items():
            if match_word in phrase and phrase in context_text:
                return ContextAnalysisResult(
                    context_score=score,
                    reason=f'Positive phrase detected: "{phrase}"',
                    is_whitelisted=True,
                )

        for phrase, score in NEGATIVE_PHRASES.items():
            if phrase in context_text:
                return ContextAnalysisResult(
                    context_score=score,
                    reason=f'Negative phrase detected: "{phrase}"',
                    is_whitelisted=False,
                )

        return None

    def _is_domain_whitelisted(self, context_words: list[str], match_word: str) -> bool:
        normalized_match_word = match_word.lower()

        for word in context_words:
            if word in self.domain_whitelists:
                return True
            if word in GAMING_POSITIVE and normalized_match_word in GAMING_ACCEPTABLE_WORDS:
                return True

        return False

    def _generate_reason(self, score: float, context_words: list[str]) -> str:
        found_positive = sorted({word for word in context_words if word in POSITIVE_INDICATORS})
        found_negative = sorted({word for word in context_words if word in NEGATIVE_INDICATORS})

        if score >= 0.7:
            details = f" (found: {', '.join(found_positive)})" if found_positive else ""
            return f"Positive context detected{details} - likely not profanity"
        if score <= 0.3:
            details = f" (found: {', '.join(found_negative)})" if found_negative else ""
            return f"Negative context detected{details} - likely profanity"
        return "Neutral context - uncertain classification"

    def _tokenize(self, text: str) -> list[str]:
        normalized = re.sub(r"[^A-Za-z0-9_\s]", " ", text.lower())
        return [word for word in normalized.split() if word]

    def _find_word_index(self, words: list[str], char_index: int) -> int:
        current_pos = 0
        for i, word in enumerate(words):
            if current_pos >= char_index:
                return max(0, i - 1)
            current_pos += len(word) + 1
        return len(words) - 1

    def _calculate_sentiment_score(
        self, context_words: list[str], match_position: int
    ) -> float:
        positive_count = 0.0
        negative_count = 0.0
        total_words = len(context_words)

        for i, word in enumerate(context_words):
            distance = abs(i - match_position)
            weight = max(0.1, 1 - (distance * 0.2))

            if word in POSITIVE_INDICATORS:
                positive_count += weight
            elif word in NEGATIVE_INDICATORS:
                negative_count += weight

        total_sentiment = positive_count + negative_count
        if total_sentiment == 0:
            return 0.5

        raw_score = positive_count / total_sentiment
        adjusted_score = raw_score

        confidence_multiplier = min(1.0, total_words / 5)
        adjusted_score = 0.5 + (adjusted_score - 0.5) * confidence_multiplier

        if any(word in {"you", "your", "u", "ur"} for word in context_words) and raw_score < 0.7:
            adjusted_score *= 0.7

        if any(
            word in {"movie", "song", "game", "book", "show", "this", "that", "it"}
            for word in context_words
        ) and raw_score > 0.3:
            adjusted_score = min(1.0, adjusted_score * 1.3)

        return max(0.0, min(1.0, adjusted_score))
